text/0000-cargo-weak-namespaced-features.md

```diff
@@ -0,0 +1,252 @@
+* The errors that versions of Cargo older than 1.51 may generate when trying to use a dependency using the new syntax can be confusing.
+* Since the `dep:` syntax no longer explicitly enables a feature of the same name, there may be some scenarios where it can be difficult to write `cfg` expressions to target exactly what the developer wants to make conditional.
+  For example:
+
+  ```toml
+  [features]
+  foo = ["dep:a", "dep:b"]
+  bar = ["dep:b"]
+  ```
+
+  Here, the developer may want to write code that is conditional on the presence of the `b` dependency.
+  With this new system, they may need to write `cfg(any(feature="foo", feature="bar"))` instead of the previously simpler syntax of `cfg(feature="b")`.
+  It is intended that in the future, syntax such as [`cfg(accessible(::b))`](https://github.com/rust-lang/rust/issues/64797) will help simplify this situation.
+  Another alternative is to rearrange the features, for example making `foo` depend on `bar` in the example above, and then use `cfg(feature="bar")` to check for the presence of `b`.
+* The new feature resolver may not emulate the old resolver behavior perfectly.
+  A large number of tests of been done to try to ensure that it works the same, but there are some unusual configurations that have not been exercised.
+  There is a moderately high risk that this may introduce unintended changes in resolver behavior or other bugs.
+
+# Rationale and alternatives
+
+* The Cargo Team considered many different variants of the syntax expressed here.
+  We feel that this hit a desired balance of expressiveness and terseness, but it certainly won't be the perfect match for everyone.
+* Instead of introducing the `dep:` syntax, Cargo could continue to keep dependencies and features in the same namespace, and instead introduce other mechanisms such as "private" features to hide optional dependencies.
+  However, several people felt that there is a conceptual benefit to splitting them into separate namespaces, and that provided a path to prevent exposing the optional dependencies and being able to use more natural feature names.
+* A new publish API could be added (endpoint `api/v2/crates/new`) to ensure that Cargo is not speaking to a registry that does not understand the new syntax.
+  This was pursued in [PR #9111](https://github.com/rust-lang/cargo/pull/9111), but it was considered not necessary.
+  [crates.io] is the only registry that can support older versions of Cargo.
+  Other registries that don't support the new syntax may reject publishing with the new syntax (if they perform validation), or they may accept it (if the don't validate), in which case it should just work.
+  The `"v"` field addition is only necessary for Cargo's older than 1.51, and most use cases of other registries are generally expected to have stricter control over which versions of Cargo are in use.
```

Comment:
As I understand it, this description of the `"v"` field is backwards. Cargo older than 1.51 will process all index entries, even if that may result in errors or unexpected behavior.

Reply:
Indeed, that was backwards. I tried to fix it, though the explanation is still a little awkward.